Add sf, big data to TADA_DataRetrieval

mbrousil opened the pull request:
Hey there KW & B,

First off, don't worry, I'm not expecting a rushed review on this! I hoped to get it out earlier this week but ran into package problems, as you know. Those aren't 100% ironed out so lmk if anything looks off to you.

There's two main changes here:

1) Addition of {sf} methods to allow users to query WQP data using {sf} objects for an aoi. I'm pretty sure that you've already reviewed the equivalent of these changes in my earlier PR to TADA_funx. So it's mostly just approving what we talked about there.

2) Folding the processes in TADA_BigDataRetrieval() into TADA_DataRetrieval(). 

**Some explanation on task 2**:

TADA_DataRetrieval() now checks the number of records and sites to be downloaded and then decides whether "big data" methods are needed. If so, that gets handed off to a new helper function, TADA_BigDataHelper(). (TADA_BigDataRetrieval() should ideally become obselete, but I've left it in here for comparison and in case we don't like my solution lol). 

I did my best to make this efficient, but the end result is that TADA_DataRetrieval() now runs more slowly on average because of the additional checks, i.e., the need to download summaries each time a query is run. At the bottom I'll add a violin plot showing the result of some `microbenchmark::microbenchmark()` runs where I downloaded **all** WQP data for Washington from 2010.

**Review request**: Could you please do a full review of the changes/additions and also:
+ let me know if you can think of any ways to speed things up
+ let me know if there's any housekeeping that I need to do before submitting to EPA?

Thank you!! And lmk if you need any other info, etc.

MRB

---

<br>

<br>


![microbench_plot](https://github.com/user-attachments/assets/005f0c36-e640-4bfc-9cb6-d461df924079)

From top to bottom, here's what each row means:

**dr_state_clean**: Original TADA_DataRetrieval() querying with `statecode = "WA"` & `applyautoclean = TRUE`
**dr_state_no_clean**: Original TADA_DataRetrieval() querying with `statecode = "WA"` & `applyautoclean = FALSE`
**bdr_state_cleaned**: Orig. TADA_**Big**DataRetrieval() querying with `statecode = "WA"` & `applyautoclean = TRUE`
**bdr_state_no_clean**: Orig. TADA_**Big**DataRetrieval() querying with `statecode = "WA"` & `applyautoclean = FALSE`
**new_fun_state_cleaned**: Updated DataRetrieval() with `statecode = "WA"` & `applyautoclean = TRUE`
**new_fun_state_no_clean**: Updated DataRetrieval() with `statecode = "WA"` & `applyautoclean = FALSE`
**new_fun_sf_cleaned**: Updated DataRetrieval() using an {sf} aoi & `applyautoclean = TRUE`
**new_fun_sf_no_clean**: Updated DataRetrieval() using an {sf} aoi & `applyautoclean = FALSE`

Note: I tried additional options where I fed in all the site_ids for the state instead of using `statecode` or `aoi_sf` but these runs just failed.

mbrousil committed:
Add sf and tribal query options to TADA_DataRetrieval

mbrousil committed:
Add TADA_TribalOptions to R/GeospatialFunctions.R

mbrousil committed:
Merge remote-tracking branch 'upstream/develop' into develop

mbrousil committed:
tribal options edits

mbrousil committed:
TADA_DR rewrite

mbrousil committed:
Helper function for large queries

mbrousil committed:
Document bigdatahelper

mbrousil committed:
update geospatial funs

mbrousil committed:
warning -> message & add dplyr::

mbrousil committed:
odds and ends

steeleb commented:
Heya! I'm going to hold off until @kathryn-willi reviews, since she's more on top of the TADA inner workings. :)

kathryn-willi commented:
Hey MB - I will be able to review this later this week. Stay tuned...

kathryn-willi commented on R/DataDiscoveryRetrieval.R:
```diff
@@ -164,171 +179,614 @@
 #'
 TADA_DataRetrieval <- function(startDate = "null",
                                endDate = "null",
+                               aoi_sf = NULL,
```
I don't remember what their reasoning is for this, but I'm guessing they will want you to replace that with "null" for consistency. Or, we full `scend and replace all! To NULL is much quicker, isn't it?

kathryn-willi commented:
1- I think that the code related to the tribal selection and using a user-supplied AOI are good to go! 🥳 But, ahead of sharing with the TADA team I suggest creating a vignette that walks through how these new arguments work and how to use them. IMO, I feel like these pieces should also be sent to TADA as their own PR without the big data stuff incorporated.

2- Seeing as the consistently fastest option of all of these is the untouched, OG dataRetrieval, I am struggling to justify using any of these other options... Is the benefit of a big data approach only the fact that it won't crap out on you? (Like, does the OG version crash on big pulls whereas the other versions don't?)

If there is still some value in a "big data" approach... Something @steeleb told me about in the office is the idea of incorporating something like this: https://www.reddit.com/r/rprogramming/comments/wb5f5t/setting_time_limit_to_execution_of_a_function_or/ Where, if the regular pull is taking a long time (like over 30 seconds maybe?), it shifts over to the big data approach automatically, and the summary data doesn't get made until AFTER the pull shifts over to the big data pull. I think I'm going to try to incorporate this into my ATTAINS functions at some point, too...

I also just remembered that TADA folks were interested in adding a progress bar for the big data pulls - since the big data pull requires an inventory ahead of download, would something like this be easy to incorporate? https://stackoverflow.com/questions/26919787/r-text-progress-bar-in-for-loop

mbrousil committed:
Add user prompts to data retrieval

mbrousil committed:
Add progress bar & change warnings to messages

mbrousil commented:
Hey @kathryn-willi and @steeleb,

I met with Cristina and Hillary in December about this and they felt that it still made sense to go forward with a single function despite the speed changes we've discussed. In essence, a simpler process for the end user is worth it.

They had a couple of related suggestions that I've implemented in the most two recent commits: 
1. Asking the user if they want to proceed given the size of the download that is expected. Include in this a note to the user about the number of rows a standard Excel file can handle since many end users will work out of Excel at some point
2. Include a progress bar

What you should see now is that there's a prompt for a "yes"/"no" from the user before downloads begin and that when the "big data" process kicks in it shows a progress bar for the % of records downloaded (separately) for sites with more & less records than `maxrecs`. 

I've also taken a small liberty here in silencing messages from `{dataRetrieval}`. It prints out GET and POST messages including the query definition and I felt that it was overwhelming the more immediately relevant information that the function intentionally prints to the user. I'm not sure if this is a change we want to keep, but to me the simplified output is easier to read.

Lastly, I changed a couple of warnings to messages because that seemed more appropriate.

steeleb commented:
Hey @mbrousil this is great. I'm going to leave the formal code review to @kathryn-willi (KW, let me know if this won't be possible for you before the EOW and we can change tack).

Given the goal of not bogging down the servers, I think this satisfies! Nice work with the interactive ask_user() function. I'm glad you silenced messages in {dataRetrieval}, I agree that's obnoxious in this function and the progress bar is a better indicator of running code.

Wheeee!

kathryn-willi commented on R/DataDiscoveryRetrieval.R:
```diff
@@ -164,171 +181,660 @@
+    }
+    
+    # Query info on available data
+    query_avail <- supressMessages(dataRetrieval::whatWQPdata(WQPquery))
```
```suggestion
    query_avail <- suppressMessages(dataRetrieval::whatWQPdata(WQPquery))
```

mbrousil committed:
Include progress bar internals

mbrousil committed:
Update DataDiscoveryRetrieval.R

kathryn-willi commented on R/DataDiscoveryRetrieval.R:
```diff
@@ -383,30 +889,155 @@ TADA_ReadWQPWebServices <- function(webservice) {
+  }
+  
+  
+  df_out <- bind_rows(df_small, df_big)
```
```suggestion
  df_out <- dplyr::bind_rows(df_small, df_big)
```

mbrousil committed:
Apply suggestions from code review

Co-authored-by: Katie/Ryn Willi (she/her) <49761053+kathryn-willi@users.noreply.github.com>

mbrousil commented on R/DataDiscoveryRetrieval.R:
```diff
@@ -164,171 +181,660 @@
+        # Get site metadata
+        sites.DR <- clipped_sites_sf %>%
+          as_tibble() %>%
+          select(-geometry)
```
```suggestion
          dplyr::select(-geometry)
```

mbrousil commented on R/DataDiscoveryRetrieval.R:
```diff
@@ -164,171 +181,660 @@
+        # Get site metadata
+        sites.DR <- clipped_sites_sf %>%
+          as_tibble() %>%
+          select(-geometry)
```
```suggestion
          dplyr::select(-geometry)
```

mbrousil committed:
Apply suggestions from code review

mbrousil committed:
Update DataDiscoveryRetrieval.R

kathryn-willi commented:
Looks good based on your additions now, MB! I say this is good to merge.
